test/artifact_test.js

```diff
@@ -227,18 +228,79 @@ suite('Artifacts', function() {
       debug('Fetching artifact from: %s', artifactUrl);
       let artifact = await getWithoutRedirecting(artifactUrl);
 
-      let expectedUrl = 
+      let expectedUrl =
         `https://test-bucket-for-any-garbage.s3-us-west-2.amazonaws.com/${taskId}/0/public/singlepart.dat`;
       assume(artifact.headers).has.property('location', expectedUrl);
 
       await verifyDownload(artifact.headers.location, bigfilehash, bigfilesize);
 
     });
 
+    test('S3 single part complete flow, content-encoding: gzip', async () => {
+      const taskId = slugid.v4();
+      const data = crypto.randomBytes(/*12 * 1024 * 1024 +*/ 21);
+      const gzipped = zlib.gzipSync(data);
+
+      debug('### Creating task');
+      await helper.queue.createTask(taskId, taskDef);
+
+      debug('### Claiming task');
+      await helper.queue.claimTask(taskId, 0, {
+        workerGroup:    'my-worker-group',
+        workerId:       'my-worker',
+      });
+
+      debug('### Create artifact');
+      const {
+        requests,
+      } = await helper.queue.createArtifact(taskId, 0, 'public/singlepart.dat', {
+        storageType: 'blob',
+        expires: taskcluster.fromNowJSON('1 day'),
+        contentType: 'binary/octet-stream',
+        contentEncoding: 'gzip',
+        contentLength: data.length,
+        contentSha256: crypto.createHash('sha256').update(data).digest('hex'),
+        transferLength: gzipped.length,
+        transferSha256: crypto.createHash('sha256').update(gzipped).digest('hex'),
+        parts: [{
+          size: gzipped.length,
+          sha256: crypto.createHash('sha256').update(gzipped).digest('hex'),
+        }],
+      });
+
+      debug('### Put first part of artifact');
+      const {method, url, headers} = requests[0];
+      const res = await request(method, url).set(headers).send(gzipped);
+      const etag = res.headers['etag'];
+
+      debug('### Complete artifact upload');
+      await helper.queue.completeArtifact(taskId, 0, 'public/singlepart.dat', {
+        etags: [etag],
+      });
+
+      const artifactUrl = helper.queue.buildUrl(
+        helper.queue.getArtifact,
+        taskId, 0, 'public/singlepart.dat',
+      );
+      debug('### Fetching artifact from: %s', artifactUrl);
+      const res2 = await request.get(artifactUrl).responseType('blob');
+      debug('Downloaded artifact, statusCode: %s', res2.status);
+      const res2Hash = crypto.createHash('sha256').update(res2.body).digest('hex');
+      assert(res2Hash === crypto.createHash('sha256').update(data).digest('hex'));
+      assert(res2Hash === res2.headers['x-amz-meta-content-sha256']);
+      debug('Response headers: %j', res2.headers);
+
+      debug('### Downloading artifact with incorrect "Accept-Encoding"');
+      let e;
+      await request.get(artifactUrl).set('Accept-Encoding', 'identity').catch(err => e = err);
```

Comment:
Let's also test for no `Accept-Encoding` value set, since that's by far the more likely scenario in reality.

Reply:
`identity` is the same thing as setting no `accept-encoding`.

Reply:
but it should also be one extra line to account for us doing something incorrectly and handling the two different options differently.